test/unit/webgl/p5.Shader.js

```diff
@@ -176,70 +152,6 @@ suite('p5.Shader', function() {
         expectedUniforms
       );
     });
-    test('Normal Shader is set after normalMaterial()', function() {
-      myp5.normalMaterial();
-      var normalShader = myp5._renderer._getNormalShader();
-      assert(
-        normalShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader was not normal shader"
-      );
-    });
-    test('Color Shader is set after fill()', function() {
-      myp5.fill(0);
-      var colorShader = myp5._renderer._getColorShader();
-      assert(
-        colorShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader was not color shader after fill"
-      );
-    });
-    test('Shader switch between retain and immedate mode', function() {
-      myp5.fill(0);
-      myp5.box(70, 70, 70);
-      var retainShader = myp5._renderer._getColorShader();
-      assert(
-        retainShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader was not color shader after fill() and box()"
-      );
-
-      myp5.beginShape(myp5.TRIANGLES);
-      myp5.vertex(0, 25, 0);
-      myp5.vertex(-25, -25, 0);
-      myp5.vertex(25, -25, 0);
-      myp5.endShape();
-      var immediateShader = myp5._renderer._getImmediateModeShader();
-      assert(
-        immediateShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader was not immediate mode shader " +
-          'after begin/endShape()'
-      );
-
-      myp5.box(70, 70, 70);
-      assert(
-        retainShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader did not switch back to retain shader " +
-          ' to draw box() after immediate mode'
-      );
-    });
-    test('Light shader set after ambientMaterial()', function() {
-      var lightShader = myp5._renderer._getLightShader();
-
-      myp5.ambientMaterial(128);
-      assert(
-        lightShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader did not get set to light shader " +
-          'after call to ambientMaterial()'
-      );
-    });
-    test('Light shader set after specularMaterial()', function() {
-      var lightShader = myp5._renderer._getLightShader();
-
-      myp5.specularMaterial(128);
-      assert(
-        lightShader === myp5._renderer.curFillShader,
-        "_renderer's curFillShader did not get set to light shader " +
-          'after call to specularMaterial()'
-      );
-    });
```

Comment:
i'm really not sure how to do this. one large part of this PR is the fact that state is no longer held in the shader, and that the shader is selected and updated immediately before drawing. the whole concept of a currently selected shader doesn't exist any more.

Reply:
I think your logic for selecting the shader in the renderer should be abstracted into a method (or two, if a separate one is needed for immediate mode) and then you can test that it would return the appropriate value after various calls to fill(), stroke(), normalMaterial(), etc.